components/raftstore/src/store/async_io/write_router.rs

```diff
@@ -238,39 +243,56 @@ where
                 let now = Instant::now();
                 if sender.send(msg, self.last_msg_priority).is_err() {
                     // Write threads are destroyed after store threads during shutdown.
-                    panic!("{} failed to send write msg, err: disconnected", self.tag);
+                    safe_panic!("{} failed to send write msg, err: disconnected", self.tag);
                 }
                 ctx.raft_metrics()
                     .write_block_wait
                     .observe(now.saturating_elapsed_secs());
             }
             Err(TrySendError::Disconnected(_)) => {
                 // Write threads are destroyed after store threads during shutdown.
-                panic!("{} failed to send write msg, err: disconnected", self.tag);
+                safe_panic!("{} failed to send write msg, err: disconnected", self.tag);
             }
         }
     }
 }
 
+pub type SenderVec<EK, ER> = Vec<Sender<WriteMsg<EK, ER>>>;
+
+#[derive(Clone)]
 /// Senders for asynchronous writes. There can be multiple senders, generally
 /// you should use `WriteRouter` to decide which sender to be used.
-#[derive(Clone)]
 pub struct WriteSenders<EK: KvEngine, ER: RaftEngine> {
-    write_senders: Vec<Sender<WriteMsg<EK, ER>>>,
+    senders: Tracker<SenderVec<EK, ER>>,
+    cached_senders: SenderVec<EK, ER>,
     io_reschedule_concurrent_count: Arc<AtomicUsize>,
 }
 
 impl<EK: KvEngine, ER: RaftEngine> WriteSenders<EK, ER> {
-    pub fn new(write_senders: Vec<Sender<WriteMsg<EK, ER>>>) -> Self {
+    pub fn new(senders: Arc<VersionTrack<SenderVec<EK, ER>>>) -> Self {
+        let cached_senders = senders.value().clone();
         WriteSenders {
-            write_senders,
+            senders: senders.tracker("async writers' tracker".to_owned()),
+            cached_senders,
             io_reschedule_concurrent_count: Arc::default(),
         }
     }
 
     #[inline]
     pub fn is_empty(&self) -> bool {
-        self.write_senders.is_empty()
+        self.cached_senders.is_empty()
+    }
+
+    #[inline]
+    pub fn size(&self) -> usize {
+        self.cached_senders.len()
+    }
+
+    #[inline]
+    pub fn refresh(&mut self) {
+        if let Some(senders) = self.senders.any_new() {
```

Comment:
How about adding a channel for batch system to broadcast message each raftstore thread. Then we can notify the change instead of checking everytime.

Reply:
Currently, I prefer to use `VersionTrack` to update all senders.

How about marking it as an optimization for this implementation as the extra channel should be compatible to `shutdown` of batch system and the extra notification mechanism? This can be regarded as an independent PR for batch system.